Architecture (#80)
chhoumann/p7: report/sections/Chapter 05 - Architecture/Architecture.tex

In this chapter we will describe the architecture of the entire system by first showing the overall structure followed by a short description of the individual components. 

> The platform is able to accomodate the ability for lecturers to create a syllabus that includes problemsets, exercises and test cases for the exercises. Additionally, the platform has the ability for students to do the exercises that the lecturer has provided. 

Reviewer: ```suggestion
The platform allows lecturers to create a syllabus that includes problemsets, exercises and test cases for the exercises. Additionally, students can do the exercises that the lecturer has provided. 
```

> The platform is a web application, consisting of a user-interface, business logic to handle requests, authentication and compilation of the code as well as a database to persist relevant data.

Reviewer: ```suggestion
The platform is a web application with a user-interface. It has business logic to handle user requests and authentication with roles, which means users of varying roles have different permissions. The platform can run tests against code, to check whether it satisfies given requirements. Account data, test runs, and syllabus data is persisted in a database.
```

Wanted to reformulate, as the sentence is long and slightly convoluted. And we don't compile the code.

The overall architecture for the system can be seen in figure \ref{fig:Architecture}.

\begin{figure}[H]
	\includegraphics[scale=0.6]{Architecture.PNG}
	\centering
	\caption{The architecture for our web application}
	\label{fig:Architecture}
\end{figure}

> The architecture for the system is based on a client-server structure. Additionally, we have added an API running a separate server instance to handle code execution and test, called \textit{Test runner}. As each component is introduced, in the following sections, we will elaborate on the choices made for each component, in relation to this structure.

Reviewer: ```suggestion
The architecture for the system is based on a client-server structure\todo{explain client-server structure}. Additionally, we have added an API running a separate server instance to handle code execution and test, called \textit{Test runner}. As each component is introduced, in the following sections, we will elaborate on the choices made for each component, in relation to this structure.
```

\section{Back-end system}
The back-end system has several responsibilities including routing, data retrieval and updates and orchestration of execution of code tests. The backend is the link between all the components in the system. It operates as a traditional server in a client-server architechture. 

\section{Frontend}
The frontend is the user facing component of our web application. The frontend is designed such that its user-interface and functionality is dependent on the role of the user. 
For example if a student has signed in, the student is able to see and complete exercises. In turn if a lecturer signs in, they will have access to more functionality such as creating a syllabus or exercise.

> Regardless of the roles of the user the frontend has the ability to persist data and execute code through communication with the backend system.

Reviewer: Maybe clarify that it isn't the frontend that has the ability to do this, but that it just asks the backend system to do it based on user actions/inputs from the frontend
